lib/datalog/roletester/src/role_tester.rs

```diff
@@ -0,0 +1,162 @@
+
+static LOGIN_TRAIT_HASH: u32 = 0;
+
+crepe! {
+    // Input from EDB
+    @input
+    struct HasRole(u32, u32);
+    @input
+    struct HasTrait(u32, u32, u32);
+    @input
+    struct NodeHasLabel(u32, u32, u32);
+    @input
+    struct RoleAllowsNodeLabel(u32, u32, u32);
+    @input
+    struct RoleDeniesNodeLabel(u32, u32, u32);
+    @input
+    struct RoleAllowsLogin(u32, u32);
+    @input
+    struct RoleDeniesLogin(u32, u32);
+
+    // Intermediate rules
+    struct HasAllowNodeLabel(u32, u32, u32, u32);
+    struct HasDenyNodeLabel(u32, u32, u32, u32);
+    struct HasAllowRole(u32, u32, u32, u32);
+    struct HasDenyRole(u32, u32, u32);
+    struct HasDeniedLogin(u32, u32, u32);
+
+    // Output for IDB
+    @output
+    struct HasAccess(u32, u32, u32, u32);
+    @output
+    struct DenyAccess(u32, u32, u32, u32);
+    @output
+    struct DenyLogins(u32, u32, u32);
+
+    // Intermediate rules to help determine access
+    HasAllowNodeLabel(role, node, key, value) <- RoleAllowsNodeLabel(role, key, value), NodeHasLabel(node, key, value);
+    HasDenyNodeLabel(role, node, key, value) <- RoleDeniesNodeLabel(role, key, value), NodeHasLabel(node, key, value);
+    HasAllowRole(user, login, node, role) <- HasRole(user, role), HasAllowNodeLabel(role, node, _, _), RoleAllowsLogin(role, login),
+        !RoleDeniesLogin(role, login);
+    HasAllowRole(user, login, node, role) <- HasRole(user, role), HasAllowNodeLabel(role, node, _, _), HasTrait(user, LOGIN_TRAIT_HASH, login),
+        !RoleDeniesLogin(role, login), !RoleDeniesLogin(role, LOGIN_TRAIT_HASH);
+    HasDenyRole(user, node, role) <- HasRole(user, role), HasDenyNodeLabel(role, node, _, _);
+    HasDeniedLogin(user, login, role) <- HasRole(user, role), RoleDeniesLogin(role, login);
+    HasDeniedLogin(user, login, role) <- HasRole(user, role), HasTrait(user, LOGIN_TRAIT_HASH, login), RoleDeniesLogin(role, LOGIN_TRAIT_HASH);
+
+    // HasAccess rule determines each access for a specified user, login and node
+    HasAccess(user, login, node, role) <- HasAllowRole(user, login, node, role), !HasDenyRole(user, node, _), !HasDeniedLogin(user, login, _);
+    DenyAccess(user, login, node, role) <- HasDenyRole(user, node, role), HasTrait(user, LOGIN_TRAIT_HASH, login);
+    DenyAccess(user, login, node, role) <- HasDenyRole(user, node, role), HasAllowRole(user, login, node, _);
+
+    DenyLogins(user, login, role) <- HasDeniedLogin(user, login, role);
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+struct Predicate {
+    atoms: Vec<u32>
+}
+
+#[derive(Serialize, Deserialize)]
+struct EDB {
+    has_role: Option<Vec<Predicate>>,
+	has_trait: Option<Vec<Predicate>>,
+	has_login_trait: Option<Vec<Predicate>>,
+	role_allows_login: Option<Vec<Predicate>>,
+	role_denies_login: Option<Vec<Predicate>>,
+	role_allows_node_label: Option<Vec<Predicate>>,
+	role_denies_node_label: Option<Vec<Predicate>>,
+	node_has_label: Option<Vec<Predicate>>,
+}
```

Comment:
Unless you need to distinguish between `None` and `Some(x)` where `x.len() == 0` you should not use options here but instead plain `Vec`'s. Because of how a `Vec` internally is 24 bytes this doesn't have any space repercussions since each field gets padded to 32 bytes anyway but it is detrimental to access performance. Unless initialised with a defined length or capacity a `Vec` does not allocate until the first element gets pushed anyway so this doesn't have any negative performance impact.